server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -306,9 +303,27 @@ public boolean isFullContent() {
         return httpRequest.body().isFull();
     }
 
+    /**
+     * Returns a copy of HTTP content. The copy is GC-managed and does not require reference counting.
+     * Please use {@link #releasableContent()} to avoid content copy.
+     */
     public BytesReference content() {
+        return BytesReference.copyBytes(releasableContent());
+    }
+
+    /**
+     * Returns a reference to the network buffer of HTTP content.
+     */
+    public ReleasableBytesReference releasableContent() {
         this.contentConsumed = true;
-        return httpRequest.body().asFull().bytes();
+        var bytes = httpRequest.body().asFull().bytes();
+        assert bytes.hasReferences() : AbstractRefCounted.ALREADY_CLOSED_MESSAGE;
+        if (bytes.hasReferences() == false) {
+            var e = new IllegalStateException(AbstractRefCounted.ALREADY_CLOSED_MESSAGE);
+            logger.error(e.getMessage(), e);
```

Comment:
I think we should log a message of our own. Adding the exception to the logger will already log the error message.
 
Something like:

"http releasable content accessed after release"

Reply:
Applied suggestion.